### ISSUE TITLE
`list()` should return promise of got `Response`

### ISSUE THREAD

**Issue by lsloan** [KTH/canvas-api#27]
I originally opened an issue about this in my org's project repo (see tl-its-umich-edu/canvas-course-manager-next#159) because I thought we were not using KTH/canvas-api correctly.  However, a colleague pointed out to me that your code is responsible for the response handling, so I'm opening this new issue.

When a Canvas API function is called without pagination handling via `get()`, the got `Response` object that was returned had a `statusCode`.  It seems that this `list()` method doesn't return that type of object, so that property (and others that may be useful) isn't available.

Also, it seems to me that `list()` should return a promise for a got `Response` object (or array of them) and the `toArray()` method could be called later to make an array of data from those responses.

For example, I'd expect to use it something like this…

```suggestion
      const response = await requestor.list<CanvasCourseSection>(endpoint, queryParams)
      logger.debug(`Received response with status code ${response.statusCode}`)
      const sections = response.toArray().map(s => ({
      …
```

(Apologies in advance for my typos.  TypeScript syntax is new to me.)

**Comment by lsloan** [comment]
I promise I wasn't actively looking for evidence to support this issue, but I thought I found some.

I had some buggy code like this…

```js
const endpoint = 'accounts/1/users'
const queryParams = { search_term: `${loginId}@example.edu` }
const usersAll = await this.requestor.list<CanvasUser>(endpoint).toArray()
```

Notice that I _forgot_ to use the `queryParams` argument in my call to `list()`.  It should have been this…

```js
const usersAll = await this.requestor.list<CanvasUser>(endpoint, queryParams).toArray()
```

Without the argument, the code seems to have gone into some endless loop.  After several minutes, my application got a `504 Gateway Time-out` error.

So, I've investigated a little more closely and found that if I make such a request to the Canvas API by other means, it tries to return **_ALL_** users in our Canvas instance.  Apparently, the long delay from `list()` may not have been an endless loop.  It may have been diligently requesting one after another of the 40,645 pages of users in our Canvas instance.

This situation may be slightly different than I thought initially.  Canvas isn't giving an error, but maybe the `list()` method should have timeout or response size limit options.  They could have reasonable default values that can be overridden when necessary.

**Comment by exacs** [comment]
Hello! Thanks for your feedback on our library. I will try to answer to your suggestions:

> Also, it seems to me that list() should return a promise for a got Response object (or array of them) and the toArray() method could be called later to make an array of data from those responses.

The method `list()` is not returning the got `Response` object and that is a design decision. We have made two endpoints to handle paginated requests: `listPaginated` and `list`. Let me explain you the difference of each:

- `listPaginated` fetches every "page" and returns its content. For example, if you run `listPaginated("accounts/1/courses")`, you will get every page. In this method, you are right, returning a `Response` instead of the body makes totally sense.
- `list` on the other hand, returns every element of each page, meaning that it automatically parses the response and, if the response is an array, will yield one element at a time.

In other words, I like your suggestion of returning the `Response` object but it makes sense to do it in `listPaginated` instead of `list`

---

> This situation may be slightly different than I thought initially. Canvas isn't giving an error, but maybe the list() method should have timeout or response size limit options. They could have reasonable default values that can be overridden when necessary.

As you can see, the `list` method returns an `ExtendedAsyncGenerator` which is like the vanilla JS iterator but with one extra method: `toArray`.

We had thoughts on implementing more methods to expand the iterator (like "filter", "map", "take") but we haven't found the need and time to implement those methods (or find a library that implements them).
